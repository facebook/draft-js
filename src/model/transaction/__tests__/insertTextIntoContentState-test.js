/**
 * Copyright (c) 2013-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @emails oncall+ui_infra
 */

'use strict';

jest.autoMockOff();

const CharacterMetadata = require('CharacterMetadata');
const Immutable = require('immutable');
const insertTextIntoContentState = require('insertTextIntoContentState');
const getSampleStateForTesting = require('getSampleStateForTesting');
const {BOLD} = require('SampleDraftInlineStyle');

describe('insertTextIntoContentState', () => {
  const sample = getSampleStateForTesting();
  const content = sample.contentState;
  const selection = sample.selectionState;

  const EMPTY = CharacterMetadata.EMPTY;

  const block = content.getBlockMap().first();

  it('must throw if selection is not collapsed', () => {
    const target = selection.set('focusOffset', 2);
    expect(() => {
      insertTextIntoContentState(content, target, 'hey', EMPTY);
    }).toThrow();
  });

  it('must return early if no text is provided', () => {
    const modified = insertTextIntoContentState(content, selection, '', EMPTY);
    expect(modified).toBe(content);
  });

  it('must insert at the start', () => {
    const character = CharacterMetadata.create({style: BOLD});
    const modified = insertTextIntoContentState(
      content,
      selection,
      'xx',
      character
    );

    const newBlock = modified.getBlockMap().first();
    expect(newBlock.getText().slice(0, 2)).toBe('xx');
    expect(newBlock.getText().slice(2)).toBe(block.getText());

    expect(
      Immutable.is(
        Immutable.List.of(
          character, character, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY
        ),
        newBlock.getCharacterList()
      )
    ).toBe(true);
  });

  it('must insert within block', () => {
    const target = selection.merge({
      focusOffset: 2,
      anchorOffset: 2,
      isBackward: false,
    });
    const character = CharacterMetadata.create({style: BOLD});
    const modified = insertTextIntoContentState(content, target, 'xx', character);
    const newBlock = modified.getBlockMap().first();

    expect(newBlock.getText().slice(2, 4)).toBe('xx');
    expect(newBlock.getText().slice(0, 2)).toBe(block.getText().slice(0, 2));
    expect(newBlock.getText().slice(4, 7)).toBe(block.getText().slice(2, 5));

    expect(
      Immutable.is(
        newBlock.getCharacterList(),
        Immutable.List([
          EMPTY, EMPTY, character, character, EMPTY, EMPTY, EMPTY,
        ])
      )
    ).toBe(true);
  });

  it('must insert at the end', () => {
    const target = selection.merge({
      focusOffset: block.getLength(),
      anchorOffset: block.getLength(),
      isBackward: false,
    });

    const character = CharacterMetadata.create({style: BOLD});
    const modified = insertTextIntoContentState(content, target, 'xx', character);

    const newBlock = modified.getBlockMap().first();
    expect(newBlock.getText().slice(5, 7)).toBe('xx');
    expect(newBlock.getText().slice(0, 5)).toBe(block.getText());

    expect(
      Immutable.is(
        newBlock.getCharacterList(),
        Immutable.List([
          EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, character, character,
        ])
      )
    ).toBe(true);
  });
});
