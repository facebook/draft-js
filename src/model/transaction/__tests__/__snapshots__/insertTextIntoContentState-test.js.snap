// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must insert at the end 1`] = `
Immutable.OrderedMap {
  "a": Immutable.Record {
    "key": "a",
    "type": "unstyled",
    "text": "Alphaxx",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "b": Immutable.Record {
    "key": "b",
    "type": "unordered-list-item",
    "text": "Bravo",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "c": Immutable.Record {
    "key": "c",
    "type": "code-block",
    "text": "Test",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "d": Immutable.Record {
    "key": "d",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "e": Immutable.Record {
    "key": "e",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "f": Immutable.Record {
    "key": "f",
    "type": "blockquote",
    "text": "Charlie",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
}
`;

exports[`must insert at the start 1`] = `
Immutable.OrderedMap {
  "a": Immutable.Record {
    "key": "a",
    "type": "unstyled",
    "text": "xxAlpha",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "b": Immutable.Record {
    "key": "b",
    "type": "unordered-list-item",
    "text": "Bravo",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "c": Immutable.Record {
    "key": "c",
    "type": "code-block",
    "text": "Test",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "d": Immutable.Record {
    "key": "d",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "e": Immutable.Record {
    "key": "e",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "f": Immutable.Record {
    "key": "f",
    "type": "blockquote",
    "text": "Charlie",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
}
`;

exports[`must insert within block 1`] = `
Immutable.OrderedMap {
  "a": Immutable.Record {
    "key": "a",
    "type": "unstyled",
    "text": "Alxxpha",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "b": Immutable.Record {
    "key": "b",
    "type": "unordered-list-item",
    "text": "Bravo",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "c": Immutable.Record {
    "key": "c",
    "type": "code-block",
    "text": "Test",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "d": Immutable.Record {
    "key": "d",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "e": Immutable.Record {
    "key": "e",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "f": Immutable.Record {
    "key": "f",
    "type": "blockquote",
    "text": "Charlie",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
}
`;

exports[`must return early if no text is provided 1`] = `
Immutable.OrderedMap {
  "a": Immutable.Record {
    "key": "a",
    "type": "unstyled",
    "text": "Alpha",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "b": Immutable.Record {
    "key": "b",
    "type": "unordered-list-item",
    "text": "Bravo",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "BOLD",
        ],
        "entity": "1",
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "c": Immutable.Record {
    "key": "c",
    "type": "code-block",
    "text": "Test",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "d": Immutable.Record {
    "key": "d",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "e": Immutable.Record {
    "key": "e",
    "type": "code-block",
    "text": "",
    "characterList": Immutable.List [],
    "depth": 0,
    "data": Immutable.Map {},
  },
  "f": Immutable.Record {
    "key": "f",
    "type": "blockquote",
    "text": "Charlie",
    "characterList": Immutable.List [
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
      Immutable.Record {
        "style": Immutable.OrderedSet [
          "ITALIC",
        ],
        "entity": null,
      },
    ],
    "depth": 0,
    "data": Immutable.Map {},
  },
}
`;
